core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -911,13 +913,33 @@ private[spark] class TaskSchedulerImpl(
     }
   }
 
-  override def executorDecommission(executorId: String): Unit = {
+  override def executorDecommission(executorId: String,
+                                    decommissionInfo: DecommissionInfo): Unit = {
+    synchronized {
+      if (!executorsPendingDecommission.contains(executorId)) {
+        executorsPendingDecommission(executorId) = decommissionInfo
+      }
+    }
     rootPool.executorDecommission(executorId)
     backend.reviveOffers()
   }
 
-  override def executorLost(executorId: String, reason: ExecutorLossReason): Unit = {
+  override def getExecutorDecommissionInfo(executorId: String): Option[DecommissionInfo] =
+    synchronized {
+   executorsPendingDecommission.get(executorId)
+  }
+
+  override def executorLost(executorId: String, givenReason: ExecutorLossReason): Unit = {
     var failedExecutor: Option[String] = None
+    val executorLostWithShuffleLost = getExecutorDecommissionInfo(executorId).map(
+      _.isShuffleLost(sc.env.blockManager.externalShuffleServiceEnabled))
```

Comment:
nit:
```suggestion
    val executorLostWithShuffleLost = getExecutorDecommissionInfo(executorId)
      .exists(_.isShuffleLost(sc.env.blockManager.externalShuffleServiceEnabled))
```

Then, we don't need `if executorLostWithShuffleLost.isDefined`.

Reply:
Neat !!!